File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/oid4vp/VpTokenBuilder.kt
package my.ssdid.sdk.domain.oid4vp

import my.ssdid.sdk.domain.sdjwt.KeyBindingJwt
import my.ssdid.sdk.domain.sdjwt.SdJwtParser
import my.ssdid.sdk.domain.sdjwt.StoredSdJwtVc

/**
 * Assembles SD-JWT VP tokens with selective disclosure and Key Binding JWT.
 *
 * Given a stored SD-JWT VC, selected claims, and verifier parameters,
 * builds a presentation token: issuerJwt~disclosure1~...~disclosureN~kbJwt
 */
object VpTokenBuilder {

    /**
     * Builds a VP token from a stored SD-JWT VC with selective disclosure.
     *
     * @param storedSdJwtVc The stored credential
     * @param selectedClaims Claim names the holder consents to disclose
     * @param audience The verifier's identifier (aud claim in KB-JWT)
     * @param nonce Verifier-provided nonce
     * @param algorithm JWA algorithm name (e.g., "EdDSA", "ES256")
     * @param signer Function that signs data with the holder's private key
     * @param issuedAt Unix timestamp for the KB-JWT (defaults to current time)
     * @return The assembled VP token string
     */
    fun build(
        storedSdJwtVc: StoredSdJwtVc,
        selectedClaims: List<String>,
        audience: String,
        nonce: String,
        algorithm: String,
        signer: (ByteArray) -> ByteArray,
        issuedAt: Long = System.currentTimeMillis() / 1000
    ): String {
        val parsed = SdJwtParser.parse(storedSdJwtVc.compact)

        val selectedDisclosures = parsed.disclosures.filter { it.claimName in selectedClaims }

        val resolvedNames = selectedDisclosures.map { it.claimName }.toSet()
        val missing = selectedClaims.toSet() - resolvedNames
        require(missing.isEmpty()) {
            "Requested claims not found in credential disclosures: $missing"
        }

        // Build the SD-JWT with selected disclosures (without KB-JWT)
        val sdJwtWithDisclosures = buildString {
            append(parsed.issuerJwt)
            append("~")
            for (disc in selectedDisclosures) {
                append(disc.encoded)
                append("~")
            }
        }

        val kbJwt = KeyBindingJwt.create(
            sdJwtWithDisclosures = sdJwtWithDisclosures,
            audience = audience,
            nonce = nonce,
            algorithm = algorithm,
            signer = signer,
            issuedAt = issuedAt
        )

        return "$sdJwtWithDisclosures$kbJwt"
    }
}

File: sdk/ios/SsdidCore/Sources/SsdidCore/Domain/OpenId4Vp/VpTokenBuilder.swift
import Foundation

/// Assembles SD-JWT VP tokens with selective disclosure and Key Binding JWT.
///
/// Given a stored SD-JWT VC, selected claims, and verifier parameters,
/// builds a presentation token: issuerJwt~disclosure1~...~disclosureN~kbJwt
public enum VpTokenBuilder {

    /// Builds a VP token from a stored SD-JWT VC with selective disclosure.
    ///
    /// - Parameters:
    ///   - storedSdJwtVc: The stored credential
    ///   - selectedClaims: Claim names the holder consents to disclose
    ///   - audience: The verifier's identifier (aud claim in KB-JWT)
    ///   - nonce: Verifier-provided nonce
    ///   - algorithm: JWA algorithm name (e.g., "EdDSA", "ES256")
    ///   - signer: Function that signs data with the holder's private key
    ///   - issuedAt: Unix timestamp for the KB-JWT (defaults to current time)
    /// - Returns: The assembled VP token string
    public static func build(
        storedSdJwtVc: StoredSdJwtVc,
        selectedClaims: [String],
        audience: String,
        nonce: String,
        algorithm: String,
        signer: (Data) -> Data,
        issuedAt: Int = Int(Date().timeIntervalSince1970)
    ) throws -> String {
        let parsed = try SdJwtParser.parse(storedSdJwtVc.compact)

        let selectedDisclosures = parsed.disclosures.filter { selectedClaims.contains($0.claimName) }

        let resolvedNames = Set(selectedDisclosures.map { $0.claimName })
        let missing = Set(selectedClaims).subtracting(resolvedNames)
        guard missing.isEmpty else {
            throw OpenId4VpError.invalidRequest(
                "Requested claims not found in credential disclosures: \(missing)"
            )
        }

        // Build the SD-JWT with selected disclosures (without KB-JWT)
        var sdJwtWithDisclosures = parsed.issuerJwt + "~"
        for disc in selectedDisclosures {
            sdJwtWithDisclosures += disc.encoded + "~"
        }

        let kbJwt = try KeyBindingJwt.create(
            sdJwtWithDisclosures: sdJwtWithDisclosures,
            audience: audience,
            nonce: nonce,
            algorithm: algorithm,
            signer: signer,
            issuedAt: issuedAt
        )

        return sdJwtWithDisclosures + kbJwt
    }
}
